pq-crypto/bike_r3/gf2x_mul.c

```diff
@@ -154,39 +66,48 @@ _INLINE_ void karatzuba(OUT uint64_t *c,
   // so we can skip the remaining two multiplications
   if(qwords_len > half_qw_len) {
     // Compute a_hi*b_hi and store the result in (c3|c2)
-    karatzuba(c2, a_hi, b_hi, qwords_len - half_qw_len, half_qw_len, sec_buf);
+    karatzuba(c2, a_hi, b_hi, qwords_len - half_qw_len, half_qw_len, sec_buf,
+              ctx);
 
     // Compute alah = (a_lo + a_hi) and blbh = (b_lo + b_hi)
-    karatzuba_add1(alah, blbh, a, b, half_qw_len);
+    ctx->karatzuba_add1(alah, blbh, a, b, half_qw_len);
 
     // Compute (c1 + c2) and store the result in tmp
-    karatzuba_add2(tmp, c1, c2, half_qw_len);
+    ctx->karatzuba_add2(tmp, c1, c2, half_qw_len);
 
     // Compute alah*blbh and store the result in (c2|c1)
-    karatzuba(c1, alah, blbh, half_qw_len, half_qw_len, sec_buf);
+    karatzuba(c1, alah, blbh, half_qw_len, half_qw_len, sec_buf, ctx);
 
     // Add (tmp|tmp) and (c3|c0) to (c2|c1)
-    karatzuba_add3(c0, tmp, half_qw_len);
+    ctx->karatzuba_add3(c0, tmp, half_qw_len);
   }
 }
 
-void gf2x_mod_mul(OUT pad_r_t *c, IN const pad_r_t *a, IN const pad_r_t *b)
+void gf2x_mod_mul_with_ctx(OUT pad_r_t *c,
+                           IN const pad_r_t *a,
+                           IN const pad_r_t *b,
+                           IN const gf2x_ctx *ctx)
 {
   bike_static_assert((R_PADDED_BYTES % 2 == 0), karatzuba_n_is_odd);
 
   DEFER_CLEANUP(dbl_pad_r_t t = {0}, dbl_pad_r_cleanup);
   ALIGN(ALIGN_BYTES) uint64_t secure_buffer[SECURE_BUFFER_QWORDS];
 
   karatzuba((uint64_t *)&t, (const uint64_t *)a, (const uint64_t *)b, R_QWORDS,
-            R_PADDED_QWORDS, secure_buffer);
+            R_PADDED_QWORDS, secure_buffer, ctx);
 
-  gf2x_red(c, &t);
+  ctx->red(c, &t);
 
   secure_clean((uint8_t *)secure_buffer, sizeof(secure_buffer));
 }
 
-void gf2x_mod_sqr_in_place(IN OUT pad_r_t *a, OUT dbl_pad_r_t *secure_buffer)
+void gf2x_mod_mul(OUT pad_r_t *c, IN const pad_r_t *a, IN const pad_r_t *b)
 {
-  gf2x_sqr(secure_buffer, a);
-  gf2x_red(a, secure_buffer);
+  bike_static_assert((R_PADDED_BYTES % 2 == 0), karatzuba_n_is_odd);
+
+  // Initialize gf2x methods struct
+  gf2x_ctx ctx;
```

Comment:
You should zero initialize ctx: `gf2x_ctx ctx = {0};`

Reply:
Done